src/go/k8s/controllers/cluster.redpanda.com/topic_controller.go

```diff
@@ -0,0 +1,634 @@
+	start := time.Now()
+	l := log.FromContext(ctx).WithName("TopicReconciler.Reconcile")
+
+	l.Info("Starting reconcile loop")
+
+	topic := &v1alpha1.Topic{}
+	if err := r.Client.Get(ctx, req.NamespacedName, topic); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !controllerutil.ContainsFinalizer(topic, FinalizerKey) {
+		patch := client.MergeFrom(topic.DeepCopy())
+		controllerutil.AddFinalizer(topic, FinalizerKey)
+		if err := r.Patch(ctx, topic, patch); err != nil {
+			l.Error(err, "unable to register finalizer")
+			return ctrl.Result{}, err
+		}
+	}
+
+	topic, result, err := r.reconcile(ctx, topic, l)
+
+	// Update status after reconciliation.
+	if updateStatusErr := r.patchTopicStatus(ctx, topic, l); updateStatusErr != nil {
+		l.Error(updateStatusErr, "unable to update topic status after reconciliation")
+		err = errors.Join(err, updateStatusErr)
+		result.Requeue = true
+	}
+
+	// Log reconciliation duration
+	durationMsg := fmt.Sprintf("reconciliation finished in %s", time.Since(start).String())
+	if result.RequeueAfter > 0 {
+		durationMsg = fmt.Sprintf("%s, next run in %s", durationMsg, result.RequeueAfter.String())
+	}
+	l.Info(durationMsg, "result", result)
+
+	if err != nil {
+		l.V(DebugLevel).Error(err, "failed to reconcile", "topic", topic)
+	} else if !topic.DeletionTimestamp.IsZero() {
+		patch := client.MergeFrom(topic.DeepCopy())
+		controllerutil.RemoveFinalizer(topic, FinalizerKey)
+		topic = v1alpha1.TopicReady(topic)
+		if err = r.Patch(ctx, topic, patch); err != nil {
+			l.Error(err, "unable to remove finalizer")
+			return ctrl.Result{}, err
+		}
+	}
+	return result, err
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *TopicReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.Topic{}).
+		Complete(r)
+}
+
+func (r *TopicReconciler) reconcile(ctx context.Context, topic *v1alpha1.Topic, l logr.Logger) (*v1alpha1.Topic, ctrl.Result, error) {
+	l = l.WithName("reconcile")
+
+	interval := time.Second * 3
+	if topic.Spec.SynchronizationInterval != nil {
+		interval = *topic.Spec.SynchronizationInterval
+	}
+
+	if topic.Status.ObservedGeneration != topic.Generation {
+		topic.Status.ObservedGeneration = topic.Generation
+		topic = v1alpha1.TopicProgressing(topic)
+		l.V(TraceLevel).Info("bump observed generation", "observed generation", topic.Generation)
+	}
+
+	kafkaClient, err := r.createKafkaClient(ctx, topic, l)
+	if err != nil {
+		return v1alpha1.TopicFailed(topic), ctrl.Result{}, err
+	}
+	defer kafkaClient.Close()
+
+	partition := int32(-1)
+	if topic.Spec.Partitions != nil {
+		partition = int32(*topic.Spec.Partitions)
+	}
+	replicationFactor := int16(-1)
+	if topic.Spec.ReplicationFactor != nil {
+		replicationFactor = int16(*topic.Spec.ReplicationFactor)
+	}
+
+	// Examine if the object is under deletion
+	if !topic.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
I think you should move this logic in the Reconcile function as the first thing you check,

Reply:
No, I can not move it as I would like to preserve that the `Reconcile` function do not have to create kafka client. Only `reconcile` function creates and manage topic. That's why in the case of topic removal/deletion reconcile function needs to be executed.